### PR TITLE
Tiny Refactor of compiler.py Pipelines

### PR DESCRIPTION
**Pull request by meawoppl**
I had a hard time telling what was going on in the pipeline(s) assembly and run stage, so I did a tiny refactor on it.  I think there is more logic than can be coalesced here, but I wanted to see if this form of change was at all agreeable to folks before stomping forward.  

Includes:
  * a new class _PipelineManager which deals with building and executing pipelines.  
  * rename/grouping of a couple of helper functions to more accurate names/places
  * Doc-string format consistancy
  * PEP8 blah blah

I want to add some documentation to this PR as well, especially around the _dispatcher class that I am having a hard time grocking.  

Feedback welcome.

### CODE REVIEW

File: numba/compiler.py [hunk 1]
```diff
@@ -163,19 +162,67 @@ def __call__(self):
...
+                        status.fail_reason = self._patch_error(msg, res.exception)
+                        break
+                # Stage OK and early escape
+                elif res.early_escape:
+                    return res.result
+            else:
+                return res.result
+
+        # TODO save all error information
+        raise CompilerError("All pipelines have failed")
```
Comment (meawoppl) [inline numba/compiler.py]: There is some pathological logic conditions on the _StageResult() object.  For example if rs.ok were True and res.early_escape were False, that could get you there, but based on perusing the _StageResult() initializers, these conditions should never arise.  

This suggests to me that there is some redundant signaling between ok and early_escape, but I don't understand well enough what they do to say for sure.  

What does the early_escap flag here really mean?

Comment (pitrou) [inline numba/compiler.py]: Apparently frontend_looplift() does its own compilation when successful, and so the object mode pipeline must be terminated early when it happens.

Comment (meawoppl) [inline numba/compiler.py]: Hmm.  Interesting.  Should I try to dissolve the exception-step in ```stage_objectmode_frontend()``` into a form more consistant with the pipeline design?  It seems funny to me to raise an exception that only gets caught in one specialized (already wrapped) function call as opposed to having a more communicative return type built into ```_Stage()``` or something similar.  

It is not completely clear to me if this is helpful, but at the very least, we can rename it something like "EarlyPipelineCompletion" which is a bit more descriptive of what it does.

Comment (pitrou) [inline numba/compiler.py]: Yes, that sounds nice, if you want to do that!

**Comment by meawoppl** [comment]
Based on conversations with @pitrou, I added some more aggressive culling of code including slicing out the ```_Stage``` and ```_StageResult``` abstractions and merging them into the ```_PipelineManager``` class.  Feedback appreciated!

File: numba/compiler.py [hunk 2]
```diff
@@ -128,54 +127,65 @@ def __repr__(self):
...
-        self.ok = ok
-        self.early_escape = early_escape
-        self._result = result
-        self._exception = exception
-
-    @property
-    def result(self):
-        assert self.ok
-        return self._result
+class _EarlyPipelineCompletion(BaseException):
```
Comment (pitrou) [inline numba/compiler.py]: This should really be an Exception, not a BaseException. BaseException is only for certain standard exceptions such as SystemExit.

Comment (meawoppl) [inline numba/compiler.py]: I will fix this after work this evening.

Comment (meawoppl) [inline numba/compiler.py]: Done.

**Comment by pitrou** [comment]
Ok, this looks good to me.

**Pull request merged**